File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Workout/Components/SetAdjustmentOverlay.swift
import SwiftUI

struct SetAdjustmentOverlay: View {
    let exercise: Exercise
    let currentSet: ExerciseSet
    let setIndex: Int
    let exerciseReadiness: ExerciseReadiness
    let weightSuggestion: WeightSuggestion?
    let averageErm: Double?
    let bodyWeight: Double?
    let loadMode: LoadModeV2
    let onDismiss: () -> Void
    let onApply: (SetAdjustmentSuggestion) -> Void

    @State private var severitySlider: Float = 0.5

    var body: some View {
        let plannedWeight: Double = {
            if loadMode == .bodyweight { return 0 }
            if let w = currentSet.weight, w > 0 { return w }
            if let sug = weightSuggestion?.suggestedWeight { return sug }
            if let erm = averageErm, let reps = currentSet.targetReps, reps > 0 {
                return erm * (1.0278 - 0.0278 * Double(reps))
            }
            return 0
        }()

        let adjustment = calculateSetAdjustment(plannedWeight: plannedWeight, readiness: exerciseReadiness, severitySlider: Double(severitySlider), averageErm: averageErm, loadMode: loadMode, bodyWeight: bodyWeight)

        let readinessColor: Color = {
            if exerciseReadiness.overallScore >= 75 { return Color(red: 0.30, green: 0.78, blue: 0.31) }
            if exerciseReadiness.overallScore >= 50 { return Color(red: 1.0, green: 0.76, blue: 0.03) }
            return Color(red: 1.0, green: 0.32, blue: 0.32)
        }()
        let readinessLabel = readinessLabel(for: exerciseReadiness.overallScore)

        let plannedModeText: String = {
            switch loadMode {
            case .lastre: return "Con Lastre"
            case .bodyweight: return "Peso Corporal"
            case .assisted: return "Asistido"
            default: return "Carga Normal"
            }
        }()
        let plannedWeightText: String = loadMode == .bodyweight ? "BW" : "\(plannedWeight.toTrimmedString())kg"

        let suggestedModeText: String = {
            switch adjustment.suggestedLoadMode {
            case .lastre: return "Con Lastre"
            case .bodyweight: return "Peso Corporal"
            case .assisted: return "Asistido"
            default: return "Carga Normal"
            }
        }()
        let suggestedWeightText: String = adjustment.suggestedLoadMode == .bodyweight ? "BW" : "\(adjustment.suggestedWeight.toTrimmedString())kg"

        VStack(spacing: 16) {
            HStack {
                VStack(alignment: .leading) {
                    Text("Ajuste por estado actual")
                        .font(.system(size: 17, weight: .bold))
                    Text("\(exercise.name) · Serie \(setIndex + 1)/\(exercise.sets.count)")
                        .font(.system(size: 12))
                        .foregroundColor(.white.opacity(0.6))
                }
                Spacer()
                Button(action: onDismiss) {
                    Image(systemName: "xmark")
                        .foregroundColor(.white.opacity(0.6))
                }
            }

            HStack(spacing: 12) {
                ZStack {
                    Circle()
                        .fill(readinessColor.opacity(0.25))
                        .frame(width: 40, height: 40)
                    Text("\(exerciseReadiness.overallScore)%")
                        .font(.system(size: 14, weight: .black))
                        .foregroundColor(readinessColor)
                }
                VStack(alignment: .leading) {
                    Text("Preparación: \(readinessLabel)")
                        .font(.system(size: 14, weight: .bold))
                    Text("Músc. \(exerciseReadiness.muscularComponent)% · SNC \(exerciseReadiness.cnsComponent)%" +
                         (exerciseReadiness.spinalWeight > 0.01 ? " · Col. \(exerciseReadiness.spinalComponent)%" : ""))
                        .font(.system(size: 11))
                        .foregroundColor(.white.opacity(0.6))
                }
                Spacer()
            }
            .padding(16)
            .background(Color(.systemGray5).opacity(0.4))
            .clipShape(RoundedRectangle(cornerRadius: 16))

            HStack(spacing: 12) {
                VStack(spacing: 4) {
                    Text(plannedModeText.uppercased())
                        .font(.system(size: 11, weight: .bold))
                        .foregroundColor(.white.opacity(0.5))
                    Text(plannedWeightText)
                        .font(.system(size: 22, weight: .black))
                }
                .frame(maxWidth: .infinity)
                .padding(14)
                .background(Color(.systemGray5))
                .clipShape(RoundedRectangle(cornerRadius: 14))

                Text("→")
                    .font(.system(size: 17, weight: .bold))
                    .foregroundColor(readinessColor)

                VStack(spacing: 4) {
                    Text(suggestedModeText.uppercased())
                        .font(.system(size: 11, weight: .bold))
                        .foregroundColor(readinessColor)
                    Text(suggestedWeightText)
                        .font(.system(size: 22, weight: .black))
                        .foregroundColor(readinessColor)
                    Text(loadMode == .assisted
                         ? "+\(Int(adjustment.suggestedWeight - plannedWeight))kg asist."
                         : "−\(Int(adjustment.reductionPercent * 100))%")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundColor(readinessColor.opacity(0.8))
                }
                .frame(maxWidth: .infinity)
                .padding(14)
                .background(readinessColor.opacity(0.15))
                .clipShape(RoundedRectangle(cornerRadius: 14))
            }

            if let erm = averageErm, erm > 0 {
                HStack(spacing: 6) {
                    Image(systemName: "info.circle")
                        .font(.system(size: 12))
                        .foregroundColor(.white.opacity(0.4))
                    Text("eRM promedio: ~\(erm.toTrimmedString())kg")
                        .font(.system(size: 11))
                        .foregroundColor(.white.opacity(0.5))
                }
            }

            VStack(alignment: .leading, spacing: 8) {
                Text("Intensidad del ajuste")
                    .font(.system(size: 14, weight: .semibold))
                Slider(value: $severitySlider, in: 0...1)
                    .tint(readinessColor)
                HStack {
                    Text("Conservador")
                        .font(.system(size: 11))
                        .foregroundColor(severitySlider <= 0.33 ? readinessColor : .white.opacity(0.4))
                    Spacer()
                    Text("Equilibrado")
                        .font(.system(size: 11))
                        .foregroundColor(severitySlider > 0.33 && severitySlider <= 0.66 ? readinessColor : .white.opacity(0.4))
                    Spacer()
                    Text("Agresivo")
                        .font(.system(size: 11))
                        .foregroundColor(severitySlider > 0.66 ? readinessColor : .white.opacity(0.4))
                }
            }

            HStack(spacing: 8) {
                Image(systemName: "info.circle")
                    .font(.system(size: 12))
                    .foregroundColor(Color(red: 0.39, green: 0.71, blue: 0.96))
                Text("El ajuste aplica solo para esta serie y sesión. Tu plan no se modifica permanentemente.")
                    .font(.system(size: 11))
                    .foregroundColor(Color(red: 0.56, green: 0.79, blue: 0.98))
            }
            .padding(12)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color(red: 0.106, green: 0.165, blue: 0.227))
            .clipShape(RoundedRectangle(cornerRadius: 10))

            HStack(spacing: 12) {
                Button(action: onDismiss) {
                    Text("Cancelar")
                        .fontWeight(.bold)
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)
                .tint(.white)
                .clipShape(RoundedRectangle(cornerRadius: 12))

                Button(action: { onApply(adjustment) }) {
                    let text: String = {
                        switch adjustment.suggestedLoadMode {
                        case .bodyweight: return "Aplicar Peso Corporal"
                        case .assisted: return "Aplicar +\(adjustment.suggestedWeight.toTrimmedString())kg Asist."
                        default: return "Aplicar \(adjustment.suggestedWeight.toTrimmedString())kg"
                        }
                    }()
                    Text(text)
                        .fontWeight(.bold)
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.borderedProminent)
                .tint(readinessColor)
                .clipShape(RoundedRectangle(cornerRadius: 12))
            }
        }
        .padding(.horizontal, 20)
        .padding(.bottom, 32)
    }
}

private func readinessLabel(for score: Int) -> String {
    if score >= 75 { return "Alta" }
    if score >= 50 { return "Media" }
    return "Baja"
}

File: android-native/app/src/main/java/com/example/kpkn/screens/workout/components/SetAdjustmentOverlay.kt
package com.example.kpkn.screens.workout.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Info
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.kpkn.data.models.*
import com.example.kpkn.domain.auge.ExerciseReadinessEngine
import com.example.kpkn.screens.workout.*
import com.example.kpkn.ui.components.KpknSheet
import kotlin.math.roundToInt

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun SetAdjustmentOverlay(
    exercise: Exercise,
    currentSet: ExerciseSet,
    setIndex: Int,
    exerciseReadiness: ExerciseReadiness,
    weightSuggestion: WeightSuggestion?,
    averageErm: Double?,
    bodyWeight: Double? = null,
    loadMode: LoadModeV2,
    onDismiss: () -> Unit,
    onApply: (SetAdjustmentSuggestion) -> Unit,
) {
    var severitySlider by remember { mutableFloatStateOf(0.5f) }
    val plannedWeight = when (loadMode) {
        LoadModeV2.BODYWEIGHT -> 0.0
        else -> currentSet.weight
            ?: weightSuggestion?.suggestedWeight
            ?: (averageErm?.let { erm ->
                if (currentSet.targetReps != null && currentSet.targetReps > 0) {
                    erm * (1.0278 - 0.0278 * currentSet.targetReps)
                } else null
            })
            ?: 0.0
    }

    val adjustment = remember(severitySlider, plannedWeight, loadMode, bodyWeight) {
        ExerciseReadinessEngine.calculateSetAdjustment(
            plannedWeight = plannedWeight,
            exerciseReadiness = exerciseReadiness,
            severitySlider = severitySlider.toDouble(),
            averageErm = averageErm,
            loadMode = loadMode,
            bodyWeight = bodyWeight,
        )
    }

    val readinessColor = when {
        exerciseReadiness.overallScore >= 75 -> Color(0xFF4CAF50)
        exerciseReadiness.overallScore >= 50 -> Color(0xFFFFC107)
        else -> Color(0xFFFF5252)
    }
    val readinessLabel = ExerciseReadinessEngine.readinessLabel(exerciseReadiness.overallScore)

    val plannedModeText = when (loadMode) {
        LoadModeV2.LOAD -> "Carga Normal"
        LoadModeV2.LASTRE -> "Con Lastre"
        LoadModeV2.BODYWEIGHT -> "Peso Corporal"
        LoadModeV2.ASSISTED -> "Asistido"
    }
    val plannedWeightText = when (loadMode) {
        LoadModeV2.BODYWEIGHT -> "BW"
        else -> "${plannedWeight.toTrimmedNumberString()}kg"
    }

    val suggestedModeText = when (adjustment.suggestedLoadMode) {
        LoadModeV2.LOAD -> "Carga Normal"
        LoadModeV2.LASTRE -> "Con Lastre"
        LoadModeV2.BODYWEIGHT -> "Peso Corporal"
        LoadModeV2.ASSISTED -> "Asistido"
    }
    val suggestedWeightText = when (adjustment.suggestedLoadMode) {
        LoadModeV2.BODYWEIGHT -> "BW"
        else -> "${adjustment.suggestedWeight.toTrimmedNumberString()}kg"
    }

    val sheetState = rememberModalBottomSheetState(
        skipPartiallyExpanded = true,
        confirmValueChange = { target -> target != SheetValue.Hidden }
    )
    KpknSheet(
        onDismissRequest = {},
        sheetState = sheetState,
        dismissible = false,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp)
                .padding(bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            // ── Header ──
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        "Ajuste por estado actual",
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.onSurface,
                    )
                    Text(
                        "${exercise.name} · Serie ${setIndex + 1}/${exercise.sets.size}",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                    )
                }
                IconButton(onClick = onDismiss) {
                    Icon(
                        Icons.Default.Close,
                        contentDescription = "Cerrar",
                        tint = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                    )
                }
            }

            // ── Info Card ──
            Card(
                shape = RoundedCornerShape(16.dp),
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
                ),
                elevation = CardDefaults.cardElevation(defaultElevation = 0.dp),
            ) {
                Row(
                    modifier = Modifier.padding(16.dp),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Box(
                        modifier = Modifier
                            .size(40.dp)
                            .clip(CircleShape)
                            .background(readinessColor.copy(alpha = 0.25f)),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(
                            "${exerciseReadiness.overallScore}%",
                            style = MaterialTheme.typography.labelLarge,
                            fontWeight = FontWeight.Black,
                            color = readinessColor,
                        )
                    }
                    Column {
                        Text(
                            "Preparación: $readinessLabel",
                            style = MaterialTheme.typography.bodyMedium,
                            fontWeight = FontWeight.Bold,
                        )
                        Text(
                            buildString {
                                append("Músc. ${exerciseReadiness.muscularComponent}%")
                                append(" · SNC ${exerciseReadiness.cnsComponent}%")
                                if (exerciseReadiness.spinalWeight > 0.01) {
                                    append(" · Col. ${exerciseReadiness.spinalComponent}%")
                                }
                            },
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                        )
                    }
                }
            }

            // ── Comparación de cargas ──
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Surface(
                    modifier = Modifier.weight(1f),
                    shape = RoundedCornerShape(14.dp),
                    color = MaterialTheme.colorScheme.surfaceVariant,
                ) {
                    Column(
                        modifier = Modifier.padding(14.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Text(
                            plannedModeText.uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
                        )
                        Spacer(Modifier.height(4.dp))
                        Text(
                            plannedWeightText,
                            style = MaterialTheme.typography.titleLarge,
                            fontWeight = FontWeight.Black,
                            color = MaterialTheme.colorScheme.onSurface,
                        )
                    }
                }

                Box(contentAlignment = Alignment.Center, modifier = Modifier.padding(horizontal = 4.dp)) {
                    Text(
                        "→",
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = readinessColor,
                    )
                }

                Surface(
                    modifier = Modifier.weight(1f),
                    shape = RoundedCornerShape(14.dp),
                    color = readinessColor.copy(alpha = 0.15f),
                ) {
                    Column(
                        modifier = Modifier.padding(14.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Text(
                            suggestedModeText.uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.Bold,
                            color = readinessColor,
                        )
                        Spacer(Modifier.height(4.dp))
                        Text(
                            suggestedWeightText,
                            style = MaterialTheme.typography.titleLarge,
                            fontWeight = FontWeight.Black,
                            color = readinessColor,
                        )
                        Text(
                            if (loadMode == LoadModeV2.ASSISTED) {
                                "+${(adjustment.suggestedWeight - plannedWeight).roundToInt()}kg asist."
                            } else {
                                "−${(adjustment.reductionPercent * 100).roundToInt()}%"
                            },
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.SemiBold,
                            color = readinessColor.copy(alpha = 0.8f),
                        )
                    }
                }
            }

            // ── Referencia eRM ──
            if (averageErm != null && averageErm > 0) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        Icons.Default.Info, null,
                        Modifier.size(12.dp),
                        tint = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                    )
                    Text(
                        "eRM promedio: ~${averageErm.toTrimmedNumberString()}kg",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
                    )
                }
            }

            // ── Slider de severidad ──
            Column {
                Text(
                    "Intensidad del ajuste",
                    style = MaterialTheme.typography.labelLarge,
                    fontWeight = FontWeight.SemiBold,
                )
                Spacer(Modifier.height(8.dp))
                Slider(
                    value = severitySlider,
                    onValueChange = { severitySlider = it },
                    modifier = Modifier.fillMaxWidth(),
                    colors = SliderDefaults.colors(
                        thumbColor = readinessColor,
                        activeTrackColor = readinessColor,
                        inactiveTrackColor = readinessColor.copy(alpha = 0.2f),
                    ),
                )
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Text(
                        "Conservador",
                        style = MaterialTheme.typography.labelSmall,
                        color = if (severitySlider <= 0.33f) readinessColor
                            else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                    )
                    Text(
                        "Equilibrado",
                        style = MaterialTheme.typography.labelSmall,
                        color = if (severitySlider in 0.34f..0.66f) readinessColor
                            else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                    )
                    Text(
                        "Agresivo",
                        style = MaterialTheme.typography.labelSmall,
                        color = if (severitySlider >= 0.67f) readinessColor
                            else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f),
                    )
                }
            }

            // ── Disclaimer ──
            Surface(
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(10.dp),
                color = Color(0xFF1B2A3A),
            ) {
                Row(
                    modifier = Modifier.padding(12.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        Icons.Default.Info, null,
                        Modifier.size(14.dp),
                        tint = Color(0xFF64B5F6),
                    )
                    Text(
                        "El ajuste aplica solo para esta serie y sesión. Tu plan no se modifica permanentemente.",
                        style = MaterialTheme.typography.labelSmall,
                        color = Color(0xFF90CAF9),
                    )
                }
            }

            // ── Botones de acción ──
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                OutlinedButton(
                    onClick = onDismiss,
                    modifier = Modifier.weight(1f),
                    shape = RoundedCornerShape(12.dp),
                ) {
                    Text("Cancelar", fontWeight = FontWeight.Bold)
                }
                Button(
                    onClick = { onApply(adjustment) },
                    modifier = Modifier.weight(1f),
                    shape = RoundedCornerShape(12.dp),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = readinessColor,
                        contentColor = Color.White,
                    ),
                ) {
                    Text(
                        when (adjustment.suggestedLoadMode) {
                            LoadModeV2.BODYWEIGHT -> "Aplicar Peso Corporal"
                            LoadModeV2.ASSISTED -> "Aplicar +${adjustment.suggestedWeight.toTrimmedNumberString()}kg Asist."
                            else -> "Aplicar ${adjustment.suggestedWeight.toTrimmedNumberString()}kg"
                        },
                        fontWeight = FontWeight.Bold,
                    )
                }
            }
        }
    }
}
